crates/store/re_dataframe2/src/query.rs

```diff
@@ -726,3 +739,681 @@ impl<'a> QueryHandle<'a> {
         std::iter::from_fn(move || self.next_row_batch())
     }
 }
+
+// ---
+
+#[cfg(test)]
+#[allow(clippy::iter_on_single_items)]
+mod tests {
+    use std::sync::Arc;
+
+    use re_chunk::{Chunk, ChunkId, RowId, TimePoint};
+    use re_chunk_store::{ChunkStore, ChunkStoreConfig, ResolvedTimeRange, TimeInt};
+    use re_log_types::{
+        build_frame_nr, build_log_time,
+        example_components::{MyColor, MyLabel, MyPoint},
+        EntityPath, Timeline,
+    };
+    use re_types_core::Loggable as _;
+
+    use crate::QueryCache;
+
+    use super::*;
+
+    // TODO(cmc): at least one basic test for every feature in `QueryExpression2`.
+    // In no particular order:
+    // * [x] filtered_index
+    // * [x] filtered_index_range
+    // * [x] view_contents
+    // * [x] selection
+    // * [ ] filtered_index_values
+    // * [ ] sampled_index_values
+    // * [ ] filtered_point_of_view
+    // * [ ] sparse_fill_strategy
+
+    // TODO(cmc): At some point I'd like to stress multi-entity queries too, but that feels less
+    // urgent considering how things are implemented (each entity lives in its own index, so it's
+    // really just more of the same).
+
+    /// All features disabled.
+    #[test]
+    fn barebones() -> anyhow::Result<()> {
+        re_log::setup_logging();
+
+        let store = create_nasty_store()?;
+        eprintln!("{store}");
+        let query_cache = QueryCache::new(&store);
+        let query_engine = QueryEngine {
+            store: &store,
+            cache: &query_cache,
+        };
+
+        let timeline = Timeline::new_sequence("frame_nr");
+        let query = QueryExpression2::new(timeline);
+        eprintln!("{query:#?}:");
+
+        let query_handle = query_engine.query(query.clone());
+        let dataframe = concatenate_record_batches(
+            query_handle.schema().clone(),
+            &query_handle.into_batch_iter().collect_vec(),
+        );
+        eprintln!("{dataframe}");
+
+        let got = format!(
+            "{:#?}",
+            dataframe.data.iter().skip(1 /* RowId */).collect_vec()
+        );
+        let expected = unindent::unindent(
+            "\
+            [
+                Int64[None, 1, 2, 3, 4, 5, 6, 7],
+                Timestamp(Nanosecond, None)[None, 1970-01-01 00:00:00.000000001, None, None, None, 1970-01-01 00:00:00.000000005, None, 1970-01-01 00:00:00.000000007],
+                ListArray[None, None, None, [2], [3], [4], None, [6]],
+                ListArray[[c], None, None, None, None, None, None, None],
+                ListArray[None, [{x: 0, y: 0}], [{x: 1, y: 1}], [{x: 2, y: 2}], [{x: 3, y: 3}], [{x: 4, y: 4}], [{x: 5, y: 5}], [{x: 8, y: 8}]],
+            ]\
+            "
+        );
```

Comment:
I find these pretty difficult to reason about without having a clear sight on the contents of store. One way would be to dump the test store in some comments. Another would be to factor the code to "emulate" a chunk view—see my other comment.